### PR TITLE
add download ability to unarchive module

### PR DESCRIPTION
**Pull request by Jmainguy**
This adds the ability to download a archive file on the remote machine, before unarchiving it.

### CODE REVIEW

File: files/unarchive.py [hunk 1]
```diff
@@ -264,11 +270,25 @@ def main():
     dest   = os.path.expanduser(module.params['dest'])
     copy   = module.params['copy']
     file_args = module.load_file_common_arguments(module.params)
+    tempdir = tempfile.mkdtemp()
 
     # did tar file arrive?
     if not os.path.exists(src):
         if copy:
             module.fail_json(msg="Source '%s' failed to transfer" % src)
+        # If copy=false, and src= contains ://, try and download the file to a temp directory.
```
Comment (bcoca) [inline files/unarchive.py]: i'm worried that with this approach you might be putting large files in memory

File: files/unarchive.py [hunk 2]
```diff
@@ -264,11 +270,25 @@ def main():
...
+        elif '://' in src:
+            package = os.path.join(tempdir, str(src.rsplit('/', 1)[1]))
+            try:
+                rsp, info = fetch_url(module, src)
+                data = rsp.read()
+                f = open(package, 'w')
```
Comment (bcoca) [inline files/unarchive.py:282]: i'm worried that with this approach you might be putting large files in memory

**Comment by Jmainguy** [comment]
Talked with Senor Coca today and decided to use the remote_tmp directory instead of creating out own tmp directory.

**Comment by Jmainguy** [comment]
Talked with team on the hangout today, change data = rsp.read() to for data in rsp.read(1024):
myfile.write(data)

this will keep stuff out of ram

**Comment by gregdek** [comment]
I'm presuming that constitutes sufficient review -- adding "P1_reviewed" tag.

**Comment by Jmainguy** [comment]
@abadger @jimi-c updated to read 1024 at a time

**Comment by abadger** [comment]
4k-64k is probably a better default but that's just a constant value.  I'll merge and then update it.  Thanks @Jmainguy !

**Comment by abadger** [comment]
Thanks @jmainguy!
  You make the community meetings worthwhile ;-)

# Merged

Hi!

This has been merged in, and will also be included in the next major release.

If you or anyone else has any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!

**Pull request merged**